joinservice/internal/certcache/certcache.go

```diff
@@ -0,0 +1,201 @@
+		ark = cacheArk
+	}
+
+	c.log.Debugf("Retrieving certificate chain from KDS")
+	kdsAsk, kdsArk, err := c.kdsClient.CertChain(signingType)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to retrieve certificate chain from KDS: %w", err)
+	}
+	if kdsAsk != nil {
+		ask = kdsAsk
+	}
+	if kdsArk != nil {
+		ark = kdsArk
+	}
+
+	askWriter := &strings.Builder{}
+	if err := pem.Encode(askWriter, &pem.Block{Type: "CERTIFICATE", Bytes: ask.Raw}); err != nil {
+		return nil, nil, fmt.Errorf("failed to encode ask: %w", err)
+	}
+	arkWriter := &strings.Builder{}
+	if err := pem.Encode(arkWriter, &pem.Block{Type: "CERTIFICATE", Bytes: ark.Raw}); err != nil {
+		return nil, nil, fmt.Errorf("failed to encode ark: %w", err)
+	}
+
+	if shouldCreateConfigMap {
+		// ConfigMap does not exist, create it.
+		c.log.Debugf("Creating certificate chain cache configmap")
+		if err := c.kubeClient.CreateConfigMap(ctx, constants.SevSnpCertCacheConfigMapName, map[string]string{
+			constants.CertCacheAskKey: askWriter.String(),
+			constants.CertCacheArkKey: arkWriter.String(),
+		}); err != nil {
+			// If the ConfigMap already exists, another JoinService instance created the certificate cache while this operation was running.
+			// Calling this function again should now retrieve the cached certificates.
+			if k8serrors.IsAlreadyExists(err) {
+				c.log.Debugf("Certificate chain cache configmap already exists, retrieving cached certificates")
+				return c.getCertChainCache(ctx)
+			}
+			return nil, nil, fmt.Errorf("failed to create certificate chain cache configmap: %w", err)
+		}
+	} else {
+		// ConfigMap already exists but either ASK or ARK are missing. Update the according value.
+		if cacheAsk == nil {
+			if err := c.kubeClient.UpdateConfigMap(ctx, constants.SevSnpCertCacheConfigMapName,
+				constants.CertCacheAskKey, askWriter.String()); err != nil {
+				return nil, nil, fmt.Errorf("failed to update ASK in certificate chain cache configmap: %w", err)
+			}
+		}
+		if cacheArk == nil {
+			if err := c.kubeClient.UpdateConfigMap(ctx, constants.SevSnpCertCacheConfigMapName,
+				constants.CertCacheArkKey, arkWriter.String()); err != nil {
+				return nil, nil, fmt.Errorf("failed to update ARK in certificate chain cache configmap: %w", err)
+			}
+		}
+	}
+
+	return ask, ark, nil
+}
+
+// getCertChainCache returns the cached ASK and ARK certificate, if available. If either of the keys
+// is not present in the configmap, no error is returned.
+func (c *Client) getCertChainCache(ctx context.Context) (ask, ark *x509.Certificate, err error) {
+	c.log.Debugf("Retrieving certificate chain from cache")
+	askRaw, err := c.kubeClient.GetConfigMapData(ctx, constants.SevSnpCertCacheConfigMapName, constants.CertCacheAskKey)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get ASK: %w", err)
+	}
+	if askRaw != "" {
+		c.log.Debugf("ASK cache hit")
+		askBlock, _ := pem.Decode([]byte(askRaw))
+		ask, err = x509.ParseCertificate(askBlock.Bytes)
```

Comment:
```suggestion
		ask, err = crypto.PemToX509Cert([]byte(askRaw))
```
pem.Decode returns nil on error, which needs to be catched. And we have a util func for this.